rfcs/0116-artifact-service.md

```diff
@@ -0,0 +1,196 @@
+  using 300-series redirects where possible
+* Where content-negotiation is impossible, the service will block creation of
+  artifacts when incoming requests aren't compatible with required response
+* We will support uploading artifacts to different regions so that we can
+
+## API
+This API is up for debate and is only an illustrative starting point.  
+
+### Generalised request format
+There will be a concept of a generalised HTTP request.  This will be an object shaped like
+```javascript
+{
+  method: 'GET',
+  url: 'https://www.example.com',
+  headers: {
+    'user-agent': 'taskcluster',
+  }
+}
+```
+
+* The body will not be included, but can be described through HTTP headers in the request
+* Duplicated HTTP headers will be unsupported (i.e. key -> string not key -> list of strings)
+* All query string, path, fragements are contained in a fully formed `url` property
+
+### Creation
+
+```
+PUT /artifacts/:name
+```
+
+This endpoint will take an origin and name.  It will return a `200` with a
+document which contains a list of generalised requests which could be run to
+complete the actual upload.  Please see the queue's current `blob` storage type
+for an example of this pattern.  The body of the request will contain metadata
+about the artifact in the shape:
+
+```javascript
+{
+  origin: '10.10.10.10' || 's3_us-west-2',
+  contentType: 'application/json',
+  contentLength: 1234,
+  contentSha256: 'abcdef1234',
+  transferLength: 1234,
+  transferSha256: 'abcdef1234',
+  contentEncdoing: 
+  expiration: new Date(),
+}
+```
+
+The `origin` parameter will either be an IPv4 address, IPv6 address, or an
+identifier.  This parameter will specify the source location of the ultimately
+originating request.  This means the origin of the request to the Queue from
+the worker and not the request to the artifact service.  If the origin is an
+IPv4 or IPv6, a mapping of the address and IP will occur to find the backing
+storage.  If the origin is not an IPv4 or IPv6 address and is identical to a
+set of known and mappable identifiers (e.g. `s3_us-west-2`), then that will be
+used to map the token to find the backing storage.
+
+### Deletetion
+
+```
+DELETE /artifacts/:name
+DELETE /caches/artifacts/:name
+```
+
+Delete all copies of the named Artifact if using first endpoint.  The second
+endpoint would attempt delete all of the cached copies in the various
+locations.  This cache purging must return a `202 Accepted` response and not
+`200/204` due to the nature of how caches work in the real world.
+
+### Retreival
+
+```
+GET /artifacts/:name[?max_time=30]
+```
+
+This endpoint returns a 302 redirect to the location of the Artifact which is
+closest to the request's IP address.  If needed, this will initiate from the
+original backing store into a different storage location.  It will wait up to a
+configurable amount of time before redirecting to the non-optimal original
+backing store.
```

Comment:
Is this always the "original" backing store, or might it be another closer source?  For example, if an artifact was uploaded to us-east-1 and is already copied to us-west-2, it would make sense to copy from us-west-2 for a us-west-1-originating request.  That sounds pretty fancy, but it's probably a good idea to leave the possibility of doing such things open.

Reply:
fair enough.  s/to the non-optimal original backing store/to a non-optimal copy/